### PR TITLE
Fix races for strMiscWarning and fLargeWork*Found, make QT runawayException use GetWarnings

### PR DESCRIPTION
**Pull request by gmaxwell**
Eliminate data races for strMiscWarning and fLargeWork*Found.

This moves all access to these datastructures through accessor functions and protects them with a lock.

and 

Make QT runawayException call GetWarnings("gui") instead of directly access strMiscWarning.

### CODE REVIEW

**Comment by gmaxwell** [comment]
I have not yet tested the QT part of this change.

**Comment by jonasschnelli** [comment]
General Concept ACK but I would prefer moving the warning function in a new file (seems to be a misuse of utils.h/c).

**Comment by MarcoFalke** [comment]
Maybe `ui_interface.*` for `GetWarnings()`?

**Comment by gmaxwell** [comment]
@MarcoFalke Can you walk me through your reason for adding a third warning context there?

**Comment by jonasschnelli** [comment]
utACK 6ed742dcbeeb16e7e5fcc42b292bc0c43f5e219c

**Comment by MarcoFalke** [comment]
My initial thinking was to put all such ui related code into `ui_interface`, but I don't think this makes sense in the long run as the original scope of `ui_interface` was different. So I am fine with putting it in the new file, though I don't see why the new header needs to go into the consensus lib... I think common or server should be enough?

**Comment by sipa** [comment]
Needs rebase.

**Comment by sipa** [comment]
Needs rebase.

File: src/warnings.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    std::string strStatusBar;
+    std::string strRPC;
+    std::string strGUI;
+    const std::string uiAlertSeperator = "<hr />";
+
+    LOCK(cs_warnings);
+
+    if (!CLIENT_VERSION_IS_RELEASE) {
+        strStatusBar = "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications";
+        strGUI = _("This is a pre-release test build - use at your own risk - do not use for mining or merchant applications");
```
Comment (gmaxwell) [inline src/warnings.cpp]: strStatusBar does that, RPC is what you get in safemode errors, effectively.

File: src/main.cpp [hunk 2]
```diff
@@ -2721,9 +2720,10 @@ void static UpdateTip(CBlockIndex *pindexNew, const CChainParams& chainParams) {
             ThresholdState state = checker.GetStateFor(pindex, chainParams.GetConsensus(), warningcache[bit]);
             if (state == THRESHOLD_ACTIVE || state == THRESHOLD_LOCKED_IN) {
                 if (state == THRESHOLD_ACTIVE) {
-                    strMiscWarning = strprintf(_("Warning: unknown new rules activated (versionbit %i)"), bit);
+                    string strWarning = strprintf(_("Warning: unknown new rules activated (versionbit %i)"), bit);
```
Comment (gmaxwell) [inline src/main.cpp]: done

**Comment by gmaxwell** [comment]
I believe I addressed the comments from @MarcoFalke

**Comment by MarcoFalke** [comment]
utACK 0b5a146cbc04949a8eb2a3f07e53c768fb0ca7dd

**Comment by gmaxwell** [comment]
Rebased for the main murder mania.

**Comment by laanwj** [comment]
Makes sense. utACK 749be01

**Pull request merged**